runtime/src/main/java/org/corfudb/protocols/logprotocol/SMRGarbageInfo.java

```diff
@@ -0,0 +1,202 @@
+ * instance whose address is the same as that of the associated log data having SMREntries.
+ *
+ * Created by Xin Li at 06/03/19.
+ */
+public class SMRGarbageInfo extends LogEntry {
+
+    /**
+     * A map to maintain information about garbage-identified SMREntry inside MultiObjectSMREntry.
+     * The key of the map is the stream id.
+     */
+    @Getter
+    private final Map<UUID, Map<Integer, SMRRecordGarbageInfo>> streamIdToGarbageMap = new ConcurrentHashMap<>();
+
+    Map<Integer, SMRRecordGarbageInfo> EMPTY_GARBAGE_MAP = new ConcurrentHashMap<>();
+
+    public SMRGarbageInfo() {
+        super(LogEntryType.SMRGARBAGE);
+    }
+
+    /**
+     * Get the garbage information about a specified SMREntry.
+     *
+     * @param streamId stream ID that the interested SMREntry is from.
+     * @param index    per-stream index of the interested SMREntry.
+     * @return if the SMREntry is identified as garbage, return the garbage information; otherwise return null.
+     */
+    public Optional<SMRRecordGarbageInfo> getGarbageInfo(UUID streamId, int index) {
+        if (!streamIdToGarbageMap.containsKey(streamId)) {
+            return Optional.empty();
+        }
+
+        return Optional.ofNullable(streamIdToGarbageMap.get(streamId).get(index));
+    }
+
+    /**
+     * Get all the garbage information of a stream.
+     *
+     * @param streamId stream ID.
+     * @return a map whose key is the per-stream index of garbage-identified SMREntries.
+     */
+    public Map<Integer, SMRRecordGarbageInfo> getAllGarbageInfo(UUID streamId) {
+        if (streamIdToGarbageMap.containsKey(streamId)) {
+            return streamIdToGarbageMap.get(streamId);
+        } else {
+            return EMPTY_GARBAGE_MAP;
+        }
+    }
+
+    /**
+     * Get all the garbage information from all streams.
+     *
+     * @return a map from streamId to another map whose key is the per-stream index of garbage-identified SMREntries.
+     */
+    public Map<UUID, Map<Integer, SMRRecordGarbageInfo>> getAllGarbageInfo() {
+        return streamIdToGarbageMap;
+    }
+
+    /**
+     * Check if there is no garbage information.
+     *
+     * @return true if there is no garbage information, false otherwise.
+     */
+    public boolean isEmpty() {
+        return streamIdToGarbageMap.isEmpty();
+    }
+
+    /**
+     * Get the total serialized size of all garbage-identified SMREntries in the associated log data.
+     *
+     * @return size in Byte.
+     */
+    public int getGarbageSize() {
+        return getGarbageSizeUpTo(Long.MAX_VALUE);
+    }
+
+    /**
+     * Get the serialized size of garbage-identified SMREntries that has detector address
+     * up to addressUpTo.
+     *
+     * @param addressUpTo detector address up to which garbage sizes are counted
+     * @return size in Byte.
+     */
+    public int getGarbageSizeUpTo(long addressUpTo) {
```

Comment:
Why `addressUpTo` is not used?

For your reference, following codes are what I used to use:
```java
public int getGarbageSizeUpTo(long addressUpTo) {
        return streamIdToGarbageMap.values().stream()
                .map(multiSmr -> multiSmr.getGarbageSizeUpTo(addressUpTo))
                .reduce(0, (a, b) -> a + b);
}
```

Reply:
done